File: packages/core-ts/src/suggestionFilter.ts
export function postprocess(
  raw: string,
  prefix: string,
  suffix: string,
  firstLineOnly: boolean,
): string | null {
  let text = raw.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  text = stripCodeFences(text);
  text = text.replace(/\s+$/, "");
  if (!text) return null;
  if (firstLineOnly) {
    text = text.split("\n")[0] ?? "";
  }
  text = removePrefixOverlap(text, prefix);
  if (!text) return null;
  if (isUseless({ suggestion: text, prefix, suffix })) return null;
  return text;
}

export function isUseless(params: {
  suggestion: string;
  prefix: string;
  suffix: string;
}): boolean {
  const trimmed = params.suggestion.trim();
  if (!trimmed) return true;
  if (params.prefix.replace(/\s+$/, "").endsWith(trimmed)) return true;
  if (params.suffix.replace(/^\s+/, "").startsWith(trimmed)) return true;
  if (duplicatesEdgeLines(params)) return true;
  if (containsRepetitivePhrase(params.suggestion)) return true;
  return false;
}

export function removePrefixOverlap(suggestion: string, prefix: string): string {
  if (!suggestion || !prefix) return suggestion;
  const max = Math.min(suggestion.length, prefix.length, 200);
  for (let n = max; n > 0; n--) {
    if (prefix.endsWith(suggestion.slice(0, n))) {
      return suggestion.slice(n);
    }
  }
  return suggestion;
}

export function shouldShowOnlyFirstLine(prefix: string, suggestion: string): boolean {
  if (!suggestion.includes("\n")) return false;
  const line = prefix.slice(prefix.lastIndexOf("\n") + 1);
  return line.length > 0 && /\S/.test(line);
}

function duplicatesEdgeLines(params: {
  suggestion: string;
  prefix: string;
  suffix: string;
}): boolean {
  const trimmed = params.suggestion.trim();
  if (!trimmed.includes("\n")) return false;
  const lines = trimmed.split("\n");
  const first = lines[0]!.trim();
  const last = lines[lines.length - 1]!.trim();
  const prefixLast = params.prefix
    .replace(/\s+$/, "")
    .slice(params.prefix.replace(/\s+$/, "").lastIndexOf("\n") + 1)
    .trim();
  const suffixFirst = params.suffix
    .replace(/^\s+/, "")
    .split("\n")[0]!
    .trim();
  if (first && prefixLast && first === prefixLast) return true;
  if (last && suffixFirst && last === suffixFirst) return true;
  return false;
}

function containsRepetitivePhrase(suggestion: string): boolean {
  const phraseLength = 20;
  const minRep = 3;
  if (suggestion.length < phraseLength * minRep) return false;
  const candidates = [suggestion, suggestion.slice(-phraseLength * minRep)];
  for (const region of candidates) {
    const phrase = region.slice(-phraseLength);
    if (!phrase.trim()) continue;
    let count = 0;
    let idx = 0;
    while (true) {
      const found = region.indexOf(phrase, idx);
      if (found < 0) break;
      count++;
      idx = found + Math.max(1, Math.floor(phrase.length / 2));
      if (count >= minRep) return true;
    }
  }
  return false;
}

function stripCodeFences(text: string): string {
  let t = text.trim();
  if (t.startsWith("```")) {
    t = t.slice(3);
    const nl = t.indexOf("\n");
    if (nl >= 0) t = t.slice(nl + 1);
    if (t.endsWith("```")) t = t.slice(0, -3);
  }
  return t;
}

File: core/src/main/kotlin/io/autocomplete/filter/SuggestionFilter.kt
package io.autocomplete.filter

/**
 * Post-process completion text and drop useless / duplicate suggestions.
 */
object SuggestionFilter {
    data class Params(
        val suggestion: String,
        val prefix: String,
        val suffix: String,
    )

    fun postprocess(
        raw: String,
        prefix: String,
        suffix: String,
        firstLineOnly: Boolean,
    ): String? {
        var text = raw.replace("\r\n", "\n").replace('\r', '\n')
        text = stripCodeFences(text)
        text = text.trimEnd()
        if (text.isEmpty()) return null
        if (firstLineOnly) {
            text = text.lineSequence().firstOrNull().orEmpty()
        }
        // Remove overlap with already typed prefix tail
        text = removePrefixOverlap(text, prefix)
        if (text.isEmpty()) return null
        if (isUseless(Params(text, prefix, suffix))) return null
        return text
    }

    fun isUseless(params: Params): Boolean {
        val trimmed = params.suggestion.trim()
        if (trimmed.isEmpty()) return true
        if (params.prefix.trimEnd().endsWith(trimmed)) return true
        if (params.suffix.trimStart().startsWith(trimmed)) return true
        if (duplicatesEdgeLines(params)) return true
        if (containsRepetitivePhrase(params.suggestion)) return true
        return false
    }

    fun removePrefixOverlap(
        suggestion: String,
        prefix: String,
    ): String {
        if (suggestion.isEmpty() || prefix.isEmpty()) return suggestion
        val max = minOf(suggestion.length, prefix.length, 200)
        var n = max
        while (n > 0) {
            if (prefix.endsWith(suggestion.substring(0, n))) {
                return suggestion.substring(n)
            }
            n--
        }
        return suggestion
    }

    fun shouldShowOnlyFirstLine(
        prefix: String,
        suggestion: String,
    ): Boolean {
        if (!suggestion.contains('\n')) return false
        val line = prefix.substringAfterLast('\n')
        // mid-line: current line already has non-whitespace content
        return line.isNotEmpty() && line.any { !it.isWhitespace() }
    }

    private fun duplicatesEdgeLines(params: Params): Boolean {
        val trimmed = params.suggestion.trim()
        if (!trimmed.contains('\n')) return false
        val lines = trimmed.split('\n')
        val first = lines.first().trim()
        val last = lines.last().trim()
        val prefixLast = params.prefix.trimEnd().substringAfterLast('\n').trim()
        val suffixFirst = params.suffix.trimStart().substringBefore('\n').trim()
        if (first.isNotEmpty() && prefixLast.isNotEmpty() && first == prefixLast) return true
        if (last.isNotEmpty() && suffixFirst.isNotEmpty() && last == suffixFirst) return true
        return false
    }

    private fun containsRepetitivePhrase(suggestion: String): Boolean {
        val phraseLength = 20
        val minRep = 3
        if (suggestion.length < phraseLength * minRep) return false
        // check both whole-string and trailing region
        val candidates = listOf(suggestion, suggestion.takeLast(phraseLength * minRep))
        for (region in candidates) {
            val phrase = region.takeLast(phraseLength)
            if (phrase.isBlank()) continue
            var count = 0
            var idx = 0
            while (true) {
                val found = region.indexOf(phrase, idx)
                if (found < 0) break
                count++
                idx = found + maxOf(1, phrase.length / 2)
                if (count >= minRep) return true
            }
        }
        return false
    }

    private fun stripCodeFences(text: String): String {
        var t = text.trim()
        if (t.startsWith("```")) {
            t = t.removePrefix("```")
            val nl = t.indexOf('\n')
            if (nl >= 0) t = t.substring(nl + 1)
            if (t.endsWith("```")) t = t.removeSuffix("```")
        }
        return t
    }
}
